Content.Shared/Prototypes/LocalizedPrototype.cs

```diff
@@ -0,0 +1,35 @@
+using System.Linq;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Prototypes;
+
+public abstract class LocalizedPrototype : IPrototype
+{
+    [IdDataField]
+    public string ID { get; } = default!;
+
+    public const string LocFormat = "{0}-{1}-{2}";
+
+    /// <summary>The localization string for the name of this prototype</summary>
+    public string NameLoc => ToLocalizationString("name");
+    /// <summary>The localized string for the name of prototype</summary>
+    public string Name => Loc.GetString(NameLoc);
+
+    /// <summary>
+    ///     Returns an Loc string using the <see cref="field"/> as the 'property'.
+    ///     Given `desc` it will return `this-prototype-PrototypeId-desc`.
+    /// </summary>
+    public string ToLocalizationString(string field)
+    {
+        // Get the ID of the proto Type
+        var type =
+            ((PrototypeAttribute?) Attribute.GetCustomAttribute(GetType(), typeof(PrototypeAttribute)))?.Type
+            ?? GetType().Name.Remove(GetType().Name.Length - 9);
+        // Lowercase the first letter
+        type = char.ToLowerInvariant(type[0]) + type[1..];
+        // Replace every uppercase letter with a dash and the lowercase letter
+        type = type.Aggregate("", (current, c) => current + (char.IsUpper(c) ? "-" + char.ToLowerInvariant(c) : c.ToString()));
```

Comment:
This performs `n` string concatenations, each concatenation having a cost of `n` itself, resulting in an `O(n²)` operation. This should use a StringBuilder instead. Considering that this function will be called hundreds or thousands of times, performance matters here.